syntax = "proto3";

package event_notifier;

service EventNotifier {
    // on client connect
    rpc OnClientConnect (OnClientConnectRequest) returns (OnClientConnectResponse) {}

    // subscribe to one event by number
    rpc SubscribeOneEventById (SubscribeOneEventByIdRequest) returns (SubscribeOneEventByIdResponse) {}

    // subscribe to events by type
    rpc SubscribeEventsByType (SubscribeEventsByTypeRequest) returns (stream SubscribeEventsByTypeResponse) {}

    // subscribe to all events
    rpc SubscribeAllEvents (SubscribeAllEventsRequest) returns (stream SubscribeAllEventsResponse) {}
}

// on client connect
message OnClientConnectRequest {
    string client_name = 1;
}

message OnClientConnectResponse {
    repeated Event events_list = 1;
    int32 num_of_events = 2;
    int32 client_id = 3;
}

// subscribe to one event by number
message SubscribeOneEventByIdRequest {
    int32 client_id = 1;
    string client_name = 2;
    string event_id = 3;
}

message SubscribeOneEventByIdResponse {
    repeated Event events_list = 1;
    Event event = 2;
    string text = 3;
}

// subscribe to events by type
message SubscribeEventsByTypeRequest {
    int32 client_id = 1;
    string client_name = 2;
    EventType event_type = 3;
}

message SubscribeEventsByTypeResponse {
    repeated Event events_list = 1;
    repeated Event subscribed_events = 2;
    string text = 3;
}

// subscribe to all events
message SubscribeAllEventsRequest {
    int32 client_id = 1;
    string client_name = 2;
}

message SubscribeAllEventsResponse {
    repeated Event events_list = 1;
    repeated Event subscribed_events = 2;
    string text = 3;
}

// event
message Event {
    string event_id = 1;
    EventType type = 2;
    map<int32, string > attendees = 3;
    int32 max_attendees = 4;
}

enum EventType {
    MEETING = 0;
    CONFERENCE = 1;
    FESTIVAL = 2;
    SPORTS = 3;
}